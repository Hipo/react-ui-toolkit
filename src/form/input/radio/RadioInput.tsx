import "./_radio-input.scss";

import React from "react";
import classNames from "classnames";

export interface RadioInputItem<Id = string, Context = any> {
  id: Id;
  content: React.ReactNode;
  inputProps: {
    htmlFor: string;
    value: string;
    name: string;
  };
  isDisabled?: boolean;
  context?: Context;
  customClassName?: string;
}

export type RadioInputSelectHandler<Id = string, Context = any> = (
  item: RadioInputItem<Id, Context>,
  event?: React.SyntheticEvent<HTMLInputElement>
) => void;

export interface RadioInputProps<Id, Context> {
  item: RadioInputItem<Id, Context>;
  onSelect: RadioInputSelectHandler<Id, Context>;
  isSelected: boolean;
  testid?: string;
  isDisabled?: boolean;
}

function RadioInput<Id = string, Context = any>({
  testid,
  item,
  onSelect,
  isSelected,
  isDisabled
}: RadioInputProps<Id, Context>) {
  const {
    inputProps: {name, value, htmlFor},
    customClassName,
    content
  } = item;
  const containerClassName = classNames("radio-input-label", customClassName, {
    "radio-input-label--is-selected": isSelected,
    "radio-input-label--is-disabled": isDisabled
  });

  return (
    <label data-testid={testid} htmlFor={htmlFor} className={containerClassName}>
      <input
        type={"radio"}
        id={htmlFor}
        name={name}
        value={value}
        className={"radio-input"}
        onChange={handleRadioButtonChange}
        checked={isSelected}
        disabled={isDisabled}
      />

      <span className={"radio-input-label__icon"} />

      {content}
    </label>
  );

  function handleRadioButtonChange(event: React.SyntheticEvent<HTMLInputElement>) {
    onSelect(item, event);
  }
}

export default RadioInput;
