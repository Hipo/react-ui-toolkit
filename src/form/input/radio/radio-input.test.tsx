import React from "react";
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../../core/utils/test/testUtils";
import RadioInput, {RadioInputProps} from "./RadioInput";

describe("<RadioInput />", () => {
  afterEach(cleanup);

  const defaultRadioInputProps: RadioInputProps = {
    testid: "radio-input",
    isSelected: false,
    onSelect: jest.fn(),
    item: {
      id: "radio-input",
      content: <p data-testid={"radio-input.content"}>{"Test"}</p>,
      inputProps: {htmlFor: "radio-input", name: "radio-input", value: "test"}
    }
  };

  it("should render correctly", () => {
    render(<RadioInput {...defaultRadioInputProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<RadioInput {...defaultRadioInputProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<RadioInput {...defaultRadioInputProps} />);

    await testA11y(container);
  });

  it("should render content correctly", () => {
    const {getByTestId} = render(<RadioInput {...defaultRadioInputProps} />);

    const radioInputContent = getByTestId("radio-input.content");

    expect(getByTestId(defaultRadioInputProps.testid!)).toContainElement(
      radioInputContent
    );
  });

  it("isSelected property should works correctly", () => {
    const {getByTestId, rerender} = render(<RadioInput {...defaultRadioInputProps} />);

    const radioInput = document.getElementsByClassName("radio-input")[0];

    expect(radioInput).not.toBeChecked();

    rerender(<RadioInput {...defaultRadioInputProps} isSelected={true} />);

    expect(radioInput).toBeChecked();
    expect(getByTestId(defaultRadioInputProps.testid!)).toHaveClass(
      "radio-input-label--is-selected"
    );
  });

  it("should call onSelect prop correctly", () => {
    render(<RadioInput {...defaultRadioInputProps} />);

    const radioInput = document.getElementsByTagName("input")[0];

    fireEvent.click(radioInput, {target: {value: "test"}});

    expect(radioInput.value).toBe("test");
    expect(defaultRadioInputProps.onSelect).toHaveBeenCalledTimes(1);
  });

  it("disabled property should works correctly", () => {
    const {getByTestId} = render(
      <RadioInput isDisabled={true} {...defaultRadioInputProps} />
    );

    const radioInput = document.getElementsByTagName("input")[0];

    expect(radioInput).toHaveAttribute("disabled");
    expect(radioInput).toBeDisabled();
    expect(getByTestId(defaultRadioInputProps.testid!)).toHaveClass(
      "radio-input-label--is-disabled"
    );
  });
});
