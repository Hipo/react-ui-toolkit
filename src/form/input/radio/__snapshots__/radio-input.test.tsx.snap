// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioInput /> should match snapshot 1`] = `
<label
  className="radio-input-label"
  data-testid="radio-input"
  htmlFor="radio-input"
>
  <input
    checked={false}
    className="radio-input"
    id="radio-input"
    name="radio-input"
    onChange={[Function]}
    type="radio"
    value="test"
  />
  <span
    className="radio-input-label__icon"
  />
  <p
    data-testid="radio-input.content"
  >
    Test
  </p>
</label>
`;
